bundle/edu.gemini.spdb.reports.collection/src/main/scala/edu/gemini/dbTools/ephemeris/TcsEphemerisExport.scala

```diff
@@ -0,0 +1,211 @@
+import edu.gemini.pot.spdb.IDBDatabaseService
+import edu.gemini.skycalc.{JulianDate, TwilightBoundedNight, Night}
+import edu.gemini.skycalc.TwilightBoundType.NAUTICAL
+import edu.gemini.spModel.core._
+import edu.gemini.spModel.core.osgi.SiteProperty
+import org.osgi.framework.BundleContext
+
+import java.io.File
+import java.nio.charset.StandardCharsets
+import java.nio.file.{Files, Paths}
+import java.security.Principal
+import java.text.SimpleDateFormat
+import java.time.Instant
+import java.util.logging.{Level, Logger}
+import java.util.{TimeZone, Date}
+
+import scalaz._
+import Scalaz._
+import scalaz.effect._
+
+/** TCS ephemeris export cron job.
+  */
+object TcsEphemerisExport {
+
+  // If specified, ephemeris files will be written here.
+  val DirectoryProp = "edu.gemini.dbTools.tcs.ephemeris.directory"
+
+  // We will request this many elements from horizons, though the actual number
+  // provided may differ.  The TCS maximum is 1440, but horizons may return a
+  // few more than requested.
+  val ElementCount  = 1430
+
+  // Text date format
+  val DateFormatString = "yyyy-MMM-dd HH:mm"
+
+  type EphemerisElement = (Coordinates, Double, Double)
+  type EphemerisMap     = Long ==>> EphemerisElement
+
+  /** Errors that may occur while exporting ephemeris data. */
+  sealed trait ExportError
+
+  /** An error that happens when working with the ODB to extract the non-
+    * sidereal observation references with their horizons ids.
+    */
+  case class OdbError(ex: Throwable) extends ExportError
+
+  /** An error that happens while working with the horizons service itself. */
+  case class HorizonsError(hid: HorizonsDesignation, e: HorizonsService2.HS2Error) extends ExportError
+
+  /** An error that happens when writing out ephemeris data for the TCS. */
+  case class WriteError(hid: HorizonsDesignation, ex: Throwable) extends ExportError
+
+  object ExportError {
+    private def reportH2Error(hid: HorizonsDesignation, h2: HS2Error): (String, Option[Throwable]) = h2 match {
+      case HorizonsService2.HorizonsError(e)   =>
+        (s"$hid: Error communicating with horizons service", Some(e))
+
+      case HorizonsService2.ParseError(_, msg) =>
+        (s"$hid: Could not parse response from horizons service: $msg", None)
+
+      case HorizonsService2.EphemerisEmpty     =>
+        (s"$hid: No response from horizons", None)
+    }
+
+    def report(e: ExportError): (String, Option[Throwable]) = e match {
+      case OdbError(ex)           =>
+        ("Error looking up nonsidereal observations in the database", Some(ex))
+
+      case HorizonsError(hid, h2) =>
+        reportH2Error(hid, h2)
+
+      case WriteError(hid, ex)    =>
+        (s"$hid: Error writing ephemeris file", Some(ex))
+    }
+
+    def logError(e: ExportError, log: Logger, prefix: String): Unit = {
+      val (msg, ex) = report(e)
+      log.log(Level.WARNING, s"$prefix$msg", ex.orNull)
+    }
+  }
+
+  type TryExport[A] = EitherT[IO, ExportError, A]
```

Comment:
I guess my point  is that these high level constructs are very difficult to follow for the non-initiated. This makes me unable to comment the code, and even less able to maintain if needed be.
I'd like that we try to level the team on the understanding of these concepts before using them so widely, or else we end up with half of the team (Or at least me :smile: ) out of sync.
The same happens in other areas, e.g. the web part, though that is more focalized on certain areas.